pedsnetdcc/db.py

```diff
@@ -0,0 +1,430 @@
+        data is fetched using cursor.fetchall() and stored in self.data. If no
+        results exist, the error is caught and None is stored instead.
+        
+        If an error occurs while executing the statement, it is caught and
+        stored in self.err and 'database error while `msg`', str(err), id_ and
+        conn_info are logged at error level. The error is not reraised.
+        The Statement object itself is returned.
+
+        If logq is given, a DictQueueHandler is created and the log messages
+        are passed through that instead of the module level logger. If resq is
+        given, the Statement object is put onto that queue after processing.
+
+        :param str conn_str: connection string for the database
+        :param Queue resq:   queue to put resulting Statement objects onto
+        :param Queue logq:   queue to put log records onto
+        :returns:            the processed Statement object itself
+        :rtype:              Statement
+        """
+
+        local_logger = self.logger
+        conn_info = get_conn_info_dict(conn.dsn)
+
+        if logq:
+            # See the DictQueueHandler docstring for the reason why the
+            # standard logging.handlers.QueueHandler can't be used here.
+            qh = DictQueueHandler(logq)
+            local_logger = logging.getLogger()
+            local_logger.setLevel(logging.DEBUG)
+            local_logger.addHandler(qh)
+
+        try:
+            with conn.cursor() as cursor:
+
+                # Execute the query.
+                msg_dict = combine_dicts({'msg': self.msg, 'sql': self.sql,
+                                          'id_': self.id_}, conn_info)
+                local_logger.debug(msg_dict)
+
+                cursor.execute(self.sql)
+
+                # Get the effected row count.
+                if cursor.rowcount in [-1, None]:
+                    self.rowcount = None
+                else:
+                    self.rowcount = cursor.rowcount
+
+                # Get the result field names.
+                if cursor.description:
+                    for field in cursor.description:
+                        self.fields.append(field[0])
+                else:
+                    self.fields = []
+
+                # Retrieve the data or, if none, set data to None.
+                try:
+                    self.data = cursor.fetchall()
+                except psycopg2.ProgrammingError as e:
+                    if e.args[0] == 'no results to fetch':
+                        self.data = None
+                    else:
+                        raise
+
+        except Exception as err:
+            self.err = err
+            msg_dict = combine_dicts({'msg': 'database error while {0}'.
+                                      format(self.msg), 'err': str(err),
+                                      'id': self.id_}, conn_info)
+            local_logger.error(msg_dict)
+
+        else:
+            self.err = None
```

Comment:
Likewise, `self`.`err` is set to none in `__init__`. It just makes me wonder if the statement object is designed so that `execute` can be run more than once. But `self`.`fields` is not (re-)initialized before use ....